Optimize //src:bazel-dev / package-bazel.sh to reduce incremental build times by 60%

jin opened the pull request:
Cuts incremental build times of the package-bazel step in //src:bazel-dev by over 60%.

On a trivial change to package-bazel.sh:

```
jingwen@jingwen1:~/bazel$ bazel build //src:bazel 
INFO: Analyzed target //src:bazel (0 packages loaded, 0 targets configured).
INFO: Found 1 target...
Target //src:bazel up-to-date:
  bazel-bin/src/bazel
INFO: Elapsed time: 5.651s, Critical Path: 5.21s
INFO: 1 process: 1 linux-sandbox.
INFO: Build completed successfully, 2 total actions
```

with `-c opt`:

```
jingwen@jingwen1:~/bazel$ bazel build //src:bazel -c opt
INFO: Analyzed target //src:bazel (0 packages loaded, 0 targets configured).
INFO: Found 1 target...
Target //src:bazel up-to-date:
  bazel-bin/src/bazel
INFO: Elapsed time: 14.577s, Critical Path: 14.18s
INFO: 1 process: 1 linux-sandbox.
INFO: Build completed successfully, 2 total actions
```

This uses the `--compilation_mode` config setting. Since the default of `-c` is `fastbuild`, this will increase the size of the regular `bazel build //src:bazel` build, but is a no-op change for `bazel build -c opt //src:bazel` release-type builds.

Alternatively, we could reuse the `DEV_BUILD` variable that is set by building `//src:bazel-dev` instead of `//src:bazel`. The implication is that folks building `//src:bazel` would have to remember to build `//src:bazel-dev` to take advantage of the faster incremental build, instead of getting it for free.

RELNOTES: Reduced the packaging time (`package-bazel.sh`) for the `//src:bazel-dev` Bazel development build target from 14s to 6s. Use `//src:bazel-dev` if you're iterating rapidly on a local Bazel changes, and use `//src:bazel --compilation_mode=opt` for release builds.

meisterT commented:
Do we run on CI with the tests with -c opt?

philwo commented:
@meisterT No, looks like we don't.

meisterT commented:
Thanks for checking, Philipp.

@jin either add `-c opt` then to CI or use the `bazel-dev` target as we do on Windows. I want to run the tests on CI with the binary that we are releasing.

jin commented:
@philwo do we build release builds with `-c opt`? I can't find this configured anywhere in our release scripts / source tree.

meisterT commented:
Yes, I added this in https://github.com/bazelbuild/bazel/commit/cc004fbe7a7cadfe89f04ab644fcfc732fe6b2e7

jin commented:
@meisterT Sounds good, I added it to bazel-dev instead. I've also updated the docs to make it more prominent that `//src:bazel-dev` is the dev build. What do you think?

jin commented:
(FYI @katre for bazel-dev.sh changes)

meisterT commented on site/docs/install-compile-source.md:
```diff
@@ -26,10 +28,11 @@ TL;DR:
-4.  Build Bazel using Bazel: `bazel build //src:bazel`
-    (or `bazel build //src:bazel-dev.exe` on Windows)
+4.  Build a development build of Bazel using Bazel: 
+     `bazel build //src:bazel-dev` (or `bazel build //src:bazel-dev.exe` on
```
super-nit: remove one space of indent
